### DISCUSSION TITLE
Is it possible to change the controller during flight?

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#682] [Q&A] [answered]
Hello.
I'm currently using the 23.02 version of firmware and Crazyflie 2.1.

As far as I know, the current default firmware is loop that enters rate control after attitude control.
I am going to implement a situation-specific adaptive control technique of the crazyflie using the gyro sensor value.
To do that, I'm trying to run with a different controller when things change in flight and then with an existing firmware controller when my attitude is restored, but when I analyzing the firmware, I'm wondering if it's a structure where the other controller can't changing during the flight.

Or if you know any theories or research findings related to agility control using Crazyflie, I'd like to be shared.

**Comment by krichardsson** [comment, marked as answer]
Hi, 
yes, you can change controller during flight, just set the [`stabilizer.controller`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#stabilizer) parameter. You can find the [mechanism that changes the controller here](https://github.com/bitcraze/crazyflie-firmware/blob/0f2f0f07d2c2463cb07e4b187f9a9a58b689d124/src/modules/src/stabilizer.c#L278-L282)

Note: When you change controller, the new controller will be initialized. It is usually not a problem, but worth keeping in mind that the new controller does not have a "history" when switched in.

**Reply by GNC-JY** [reply]
So can switching be done smoothly if I keep saving history data or running 2 controllers?

**Reply by krichardsson** [reply]
I think it will work to just change controller without any modifications, the best approach is probably to test what happens.
If there is a problem you know where to look :-)
